### PR TITLE
feat: new app to create reports based on sql queries

### PR DESCRIPTION
**Pull request by mswertz**
To test:
* create new schema (pet store includes an example)
* in address bar you will find a new 'reports' item
* alternatively, go to existing schema and then in address bar replace last bit url with /reports

Features:
* define reports using sql
* can download report as zip
* can download multiple as zip (names will be used for the csv file names)

P.S. remember, in sql you need to quote identifiers if they are not lower case. For example

```select * from "Pet"```

### CODE REVIEW

File: apps/molgenis-components/src/components/layout/Molgenis.vue [hunk 1]
```diff
@@ -66,12 +66,17 @@ export default {
         {
           label: "Schema",
           href: "schema",
-          role: "Manager",
+          role: "Viewer",
```
Comment (connoratrug) [inline apps/molgenis-components/src/components/layout/Molgenis.vue:69]: schema viewer ?

File: apps/molgenis-components/src/components/layout/Molgenis.vue [hunk 2]
```diff
@@ -66,12 +66,17 @@ export default {
...
         },
         {
           label: "Up/Download",
           href: "updownload",
-          role: "Editor",
+          role: "Viewer",
```
Comment (connoratrug) [inline apps/molgenis-components/src/components/layout/Molgenis.vue:74]: updown Viewer ?

File: apps/molgenis-components/src/components/tables/TableMolgenis.vue [hunk 3]
```diff
@@ -205,7 +205,7 @@ export default {
         }
         this.$emit("update:selection", update);
       } else {
-        this.$emit("click", key);
+        this.$emit("rowClick", key);
```
Comment (connoratrug) [inline apps/molgenis-components/src/components/tables/TableMolgenis.vue:208]: where is this change handled ?

File: apps/reports/README.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
+# tables
+
+## Project setup
+```
+npm install
```
Comment (connoratrug) [inline apps/reports/README.md:5]: yarn

File: apps/molgenis-components/src/components/layout/Molgenis.vue (hunk 2 above)
Comment (mswertz) [reply apps/molgenis-components/src/components/layout/Molgenis.vue:74]: yes, it is allowed to download, not to upload. So this is useful for a viewer also

File: apps/molgenis-components/src/components/layout/Molgenis.vue (hunk 1 above)
Comment (mswertz) [reply apps/molgenis-components/src/components/layout/Molgenis.vue:69]: Yes, schema is useful for somebody with view permission as documentation. It is not only editing of the schema

File: apps/molgenis-components/src/components/layout/Molgenis.vue (hunk 2 above)
Comment (connoratrug) [reply apps/molgenis-components/src/components/layout/Molgenis.vue:74]: check

File: apps/molgenis-components/src/components/layout/Molgenis.vue (hunk 1 above)
Comment (connoratrug) [reply apps/molgenis-components/src/components/layout/Molgenis.vue:69]: check

**Review by connoratrug — CHANGES_REQUESTED** [review]
when i create a report and click de pencil icon i get an error 

Error: Error: Exception while fetching data (/_reports) : Retrieve of report '0' failed : Transaction failed: current transaction is aborted, commands ignored until end of transaction block.: SQL [RESET search_path]; ERROR: current transaction is aborted, commands ignored until end of transaction block: {"response":{"errors":[{"message":"Error: Exception while fetching data (/_reports) : Retrieve of report '0' failed : Transaction failed: current transaction is aborted, commands ignored until end of transaction block.: SQL [RESET search_path]; ERROR: current transaction is aborted, commands ignored until end of transaction block"}],"status":400,"headers":{"map":{"connection":"keep-alive","content-type":"application/json","date":"Wed, 14 Dec 2022 15:12:22 GMT","strict-transport-security":"max-age=15724800; includeSubDomains","transfer-encoding":"chunked"}}},"request":{"query":"{_reports(id:0,limit:5,offset:0){data,count}}"}}

File: docs/molgenis/use_global_settings.md [hunk 5]
```diff
@@ -3,10 +3,14 @@
 You can find the 'admin' menu when on the MOLGENIS start page, when you view the list of databases (click the MOLGENIS
 logo to get there). Only when signed in as 'admin' user will this menu item be shown.
 
-Settings currently supported:
-
+Database settings currently supported:
 * LANDING_PAGE - to change landing page from default /apps/central to something else
 
+Schema settings currently supported:
+* menu - will be set by the menus app
```
Comment (connoratrug) [inline docs/molgenis/use_global_settings.md:10]: menus app ?

File: backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/ZipApi.java [hunk 6]
```diff
@@ -39,6 +46,10 @@ public static void create() {
     // table level operations
     final String tablePath = "/:schema/api/zip/:table"; // NOSONAR
     get(tablePath, ZipApi::getZipTable);
+
+    // query operator
+    final String reportPath = "/:schema/api/reports/zip"; // NOSONAR
+    get(reportPath, ZipApi::getZippedReports);
```
Comment (connoratrug) [inline backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/ZipApi.java:52]: no sonar ? , suggest to inline path in get

**Review by connoratrug — CHANGES_REQUESTED** [review]
Allowing the end user to execute sql prevents any filtering / blocking of certain records/column via app code. The general EMX2 design decisions should be updated, not sure if this feature would be worth it

File: apps/reports/src/components/ViewEditReport.vue [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    download(id) {
+      window.open("../api/reports/zip?id="+id, '_blank');
+    }
+  },
+  watch: {
+    page() {
+      this.run();
+    }
+  },
+  async mounted() {
```
Comment (jelmerveen) [inline apps/reports/src/components/ViewEditReport.vue:117]: no need for async

File: backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlConstants.java [hunk 8]
```diff
@@ -48,6 +48,7 @@ public class GraphqlConstants {
   public static final String MESSAGE = "message";
   public static final String TOKEN = "token";
   public static final String STATUS = "status";
+  public static final String DATA = "data";
 
   public static final String OPERATION = "operation";
   public static final String STAMP = "stamp";
```
Comment (jelmerveen) [inline backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/graphql/GraphqlConstants.java:54]: I do not know if it is possible, but all these look more suitable for an ENUM to me

**Review by jelmerveen — CHANGES_REQUESTED** [review]

File: apps/reports/README.md (hunk 4 above)
Comment (mswertz) [reply apps/reports/README.md:5]: this is error in all apps, now only fixed here

**Comment by mswertz** [comment]
> Allowing the end user to execute sql prevents any filtering / blocking of certain records/column via app code. The general EMX2 design decisions should be updated, not sure if this feature would be worth it

This is already the general design decision of EMX2. Permissions should always be enforced on database setting. That keep door open for a microservice architecture.

File: apps/molgenis-components/src/components/tables/TableSimple.vue [hunk 9]
```diff
@@ -89,7 +84,7 @@ export default {
   methods: {
     flattenObject(object) {
       let result = "";
-      if (typeof object === "object") {
+      if (object && typeof object === "object") {
```
[the remark this answers is not in the document]
Comment (mswertz) [reply apps/molgenis-components/src/components/tables/TableSimple.vue:87]: yes, that is okay.

**Comment by connoratrug** [comment]
> 
> P.S. end user cannot define the sql, only users with 'manager' permissions can define it.

a 'manager' is a 'end-user' in this context for me 

> keep door open for a microservice architecture.

That sound like a good decision, just checking that the decision was actually made (and communicated)  instead of it being a side-effect of this feature pr.

**Review by jelmerveen — APPROVED** [review]

**Pull request merged**